src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+
+        public unsafe void HandleEvents(uint events)
+        {
+            lock (_closeLock)
+            {
+                if ((events & Interop.libc.EPOLLERR) != 0)
+                {
+                    int errno;
+                    uint optLen = (uint)sizeof(int);
+                    int err = Interop.libc.getsockopt(_fileDescriptor, Interop.libc.SOL_SOCKET, Interop.libc.SO_ERROR, &errno, &optLen);
+                    if (err == -1)
+                    {
+                        // TODO: throw an appropiate exception
+                        throw new Exception(string.Format("HandleEvents getsockopt: {0}", Interop.Sys.GetLastError()));
+                    }
+
+                    // TODO: error handling
+                }
+
+                // TODO: rethink ordering of HUP/RDHUP w.r.t. IN/OUT.
+
+                if ((events & Interop.libc.EPOLLHUP) != 0)
+                {
+                    // Drain queues and unregister fd
+
+                    Queue<AcceptOrConnectOperation> acceptOrConnectQueue;
+                    Queue<TransferOperation> sendQueue;
+                    Queue<TransferOperation> receiveQueue;
+                    lock (_queueLock)
+                    {
+                        acceptOrConnectQueue = _acceptOrConnectQueue.Stop();
+                        sendQueue = _sendQueue.Stop();
+                        receiveQueue = _receiveQueue.Stop();
+
+                        if (_registeredEvents != 0)
+                        {
+                            Interop.Error errorCode;
+                            if (!_engine.TryUnregister(ref _handle, _fileDescriptor, 0, out errorCode))
+                            {
+                                if (errorCode != Interop.Error.EBADF)
+                                {
+                                    // TODO: throw an appropiate exception
+                                    throw new Exception(string.Format("HandleEvents HUP: {0}", errorCode));
+                                }
+                                _handle.Free();
+                            }
+
+                            _registeredEvents = 0;
+                        }
+
+                        // TODO: assert that queues are all empty if _registeredEvents was zero?
+
+                        Debug.Assert(!_handle.IsAllocated);
+                    }
+
+                    while (!acceptOrConnectQueue.IsEmpty)
+                    {
+                        AcceptOrConnectOperation op = acceptOrConnectQueue.Head;
+
+                        bool completed;
+
+                        var acceptOp = op as AcceptOperation;
+                        if (acceptOp != null)
+                        {
+                            completed = TryCompleteAccept(_fileDescriptor, acceptOp);
+                        }
+                        else
+                        {
+                            completed = TryCompleteConnect(_fileDescriptor, (ConnectOperation)op);
+                        }
```

Comment:
I think I've heard you say you don't like this construction, but FWIW, I prefer seeing such statements as:
```C#
bool completed = (acceptOp != null) ?
    TryCompleteAccept(_fileDescriptor, acceptOp) :
    TryCompleteConnect(_fileDescriptor, (ConnectOperation)op);
```